csrc/index_compute.cpp

```diff
@@ -3878,38 +3885,98 @@ std::pair<Val*, Val*> Index::getCpAsyncBulkGmemIndex(
 
   int64_t itemsize = dataTypeSize(gmem_tv->dtype());
 
+  // Loop through the frontier from inner to outer, handle TMA-global ID and
+  // non-TMA-global ID separately. Each TMA-global ID will be treated as a
+  // separate tensor dimension. Non-TMA-global ID will be treated as if they
+  // are merged together with respect to their contiguity.
+  //
+  // Use the following 2-state machine to track when we need to create a new
+  // dimension and when we need to merge the current non-TMA-global ID with its
+  // next dimension:
+  //
+  //                TMA-global ID
+  //               .-------------.
+  //               |             |
+  //               '-> [START] --'
+  //                   /    ^
+  //   non-TMA-global /      \ TMA-global
+  //        ID       |        |   ID
+  //                  \      /
+  //                   v    /
+  //              .-> [PENDING] --.
+  //              |               |
+  //              '---------------'
+  //              non-TMA-global ID
```

Comment:
Not sure how to interpret this state diagram.  What does the state imply? Wonder what it means when the state returns to START. 

As for the arrow, I suppose the label refers to the outer domain of the current domain, correct? The next state means the state of the domain when two domains are merged?

Reply:
Looks like the doc needs a lot more clarification. I did it in https://github.com/NVIDIA/Fuser/pull/2004/commits/989092638145aa5f2bfc7b7171d381165495ea4b, hopefully it will be more understandable. Every time when I use a state machine, I feel that it does make the code clean and neat, but it also makes the idea very difficult to explain. This happened a few times in the past... Hopefully, I answered all your questions in the updated doc.